src/main/java/com/coreos/jetcd/internal/impl/ClientImpl.java

```diff
@@ -13,178 +10,152 @@
-    this.watchClient = Suppliers.memoize(() -> new WatchImpl(this));
-  }
-
-  Pair<ManagedChannel, Optional<String>> toChannelAndToken(ManagedChannelBuilder<?> channelBuilder)
-      throws AuthFailedException, ConnectException {
-    checkNotNull(channelBuilder, "channelBuilder can't be null");
-    ManagedChannel managedChannel = channelBuilder.build();
-    Optional<String> token = generateToken(managedChannel, this.user, this.pass);
-    return new Pair<>(managedChannel, token);
   }
 
+  @Override
   public Auth getAuthClient() {
-    return authClient.get();
+    return newClient(authClient, AuthImpl::new);
   }
 
+  @Override
   public KV getKVClient() {
-    return kvClient.get();
+    return newClient(kvClient, KVImpl::new);
   }
 
+  @Override
   public Cluster getClusterClient() {
-    return clusterClient.get();
+    return newClient(clusterClient, ClusterImpl::new);
   }
 
+  @Override
   public Maintenance getMaintenanceClient() {
-    return this.maintenanceClient.get();
+    return newClient(maintenanceClient, MaintenanceImpl::new);
   }
 
+  @Override
   public Lease getLeaseClient() {
-    return this.leaseClient.get();
+    return newClient(leaseClient, LeaseImpl::new);
   }
 
+  @Override
   public Watch getWatchClient() {
-    return this.watchClient.get();
+    return newClient(watchClient, WatchImpl::new);
   }
 
-  public void close() {
-    this.getLeaseClient().close();
-    this.executorService.shutdownNow();
-    this.channel.shutdownNow();
+  @Override
+  public synchronized void close() {
+    Optional.ofNullable(authClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(kvClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(clusterClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(maintenanceClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(leaseClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(watchClient.get()).ifPresent(OperationClient::close);
+    Optional.ofNullable(managedConnection.get()).ifPresent(ManagedConnection::close);
   }
 
   /**
-   * get token from etcd with name and password.
+   * Create a managed connection.
    *
-   * @param channel channel to etcd
-   * @param name auth name
-   * @param password auth password
-   * @return authResp
+   * @return the managed connection
+   * @throws ConnectException This may be caused as network reason, wrong address
+   * @throws AuthFailedException This may be caused as wrong username or password
    */
-  private static ListenableFuture<AuthenticateResponse> authenticate(ManagedChannel channel,
-      ByteString name, ByteString password) {
-    return AuthGrpc.newFutureStub(channel).authenticate(
-        AuthenticateRequest.newBuilder()
-            .setNameBytes(name)
-            .setPasswordBytes(password)
-            .build()
-    );
+  private ManagedConnection getManagedConnection() throws AuthFailedException, ConnectException {
```

Comment:
why throw unchecked `AuthFailedException, ConnectException`? is this for java doc?